producer/awsS3.go

```diff
@@ -0,0 +1,281 @@
+	// get batchedFile from filesByStream[streamID] map
+	prod.batchedFileGuard.RLock()
+	batchedFile, fileExists := prod.filesByStream[streamID]
+	prod.batchedFileGuard.RUnlock()
+	if fileExists {
+		if rotate, err := prod.needsRotate(batchedFile, forceRotate); !rotate {
+			return batchedFile, err // ### return, already open or error ###
+		}
+	}
+
+	prod.batchedFileGuard.Lock()
+	defer prod.batchedFileGuard.Unlock()
+
+	// check again to avoid race conditions
+	if batchedFile, fileExists := prod.filesByStream[streamID]; fileExists {
+		if rotate, err := prod.needsRotate(batchedFile, forceRotate); !rotate {
+			return batchedFile, err // ### return, already open or error ###
+		}
+	}
+
+	baseFileName := prod.getBaseFileName(streamID)
+
+	// get batchedFile from files[path] and assure the file is correctly mapped
+	batchedFile, fileExists = prod.files[baseFileName]
+	if !fileExists {
+		batchedFile = prod.newBatchedWriterAssembly()
+
+		prod.filesByStream[streamID] = batchedFile
+		prod.files[baseFileName] = batchedFile
+	}
+
+	// Close existing batchedFile.writer
+	if batchedFile.HasWriter() {
+		oldAwsWriter := batchedFile.GetWriterAndUnset()
+
+		prod.Logger.Info("Rotated ", oldAwsWriter.Name(), " -> ", baseFileName)
+		go oldAwsWriter.Close() // close in subroutine for eventually compression in the background
+	}
+
+	// Update BatchedWriterAssembly writer
+	writer := awsS3.NewBatchedFileWriter(prod.s3Client, prod.bucket, prod.getFinalFileName(baseFileName), prod.Logger)
+	batchedFile.SetWriter(&writer)
+
+	return batchedFile, nil
+}
+
+func (prod *AwsS3) needsRotate(batchedFile *components.BatchedWriterAssembly, forceRotate bool) (bool, error) {
+	// run default rotation checks
+	if needUpload, err := batchedFile.NeedsRotate(prod.Rotate, forceRotate); needUpload {
+		return true, err
+	}
+
+	// check if max multipart uploads of 1000 reached
+	// @see: http://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html
+	// we use 995 to have a small buffer to the limit and need at least +1 upload part for the last flush
+	writer, ok := batchedFile.GetWriter().(awsS3.BatchedFileWriterInterface)
+	if ok && writer.GetUploadCount() > 995 {
+		prod.Logger.Debug("Rotate true: ", "upload count reached limit of 1000")
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func (prod *AwsS3) getBaseFileName(streamID core.MessageStreamID) string {
+	if prod.hasWildcard {
+		streamName := core.StreamRegistry.GetStreamName(streamID)
+		return strings.Replace(prod.fileNamePattern, "*", streamName, -1)
+	}
+
+	return prod.fileNamePattern
+}
+
+func (prod *AwsS3) getFinalFileName(baseFileName string) string {
+	fileExt := filepath.Ext(baseFileName)
+	fileName := baseFileName[:len(baseFileName)-len(fileExt)]
+
+	timestamp := time.Now().Format(prod.Rotate.Timestamp)
+	signature := fmt.Sprintf("%s_%s", fileName, timestamp)
+	maxSuffix := uint64(0)
+
+	//todo: get list from aws?
+	/*files, _ := ioutil.ReadDir(streamFile.dir)
```

Comment:
As discussed.
To avoid too many HTTP calls (e.g. multiple rotates per second) we should list the files once at start and keep a "cache" of the latest file name. As writing to the same file/folder from different instances will break, we can be pretty sure we are the only one doing this.

However - this was decided to be too much for 0.5.0, so create an issue for 0.5.x so we support this feature later (because it is a valid thing)